core/consensus/grandpa/impl/voting_round_impl.cpp

```diff
@@ -1112,98 +1122,58 @@ namespace kagome::consensus::grandpa {
+               "(total prevote weight={} < threshold={})",
                round_number_,
-               precommits_->getTotalWeight(),
+               prevotes_->getTotalWeight(),
                threshold_);
       return false;
     }
 
-    auto possible_to_finalize = [this](const VoteWeight &weight) {
-      return weight.total(precommit_equivocators_, *voter_set_) >= threshold_;
-    };
-
-    auto current_best = precommit_ghost_.value_or(prevote_ghost_.value_or([&] {
-      auto previous_round = previous_round_.lock();
-      return previous_round ? previous_round->bestFinalCandidate()
-                            : last_finalized_block_;
-    }()));
-
-    /// @see spec: Grandpa-Ghost
-    auto new_precommit_ghost =
-        precommit_graph_->findGhost(current_best, possible_to_finalize);
-
-    if (not new_precommit_ghost.has_value()) {
-      new_precommit_ghost = current_best;
-
-      SL_TRACE(logger_,
-               "Round #{}: updatePrecommitGhost <- not finalizable (best #{})",
-               round_number_,
-               current_best.number);
-    } else {
-      SL_TRACE(logger_,
-               "Round #{}: updatePrecommitGhost <- finalizable (#{}, best #{})",
-               round_number_,
-               new_precommit_ghost.value().number,
-               current_best.number);
-
-      finalized_ = new_precommit_ghost.value();
-    }
-
-    bool changed = new_precommit_ghost != precommit_ghost_;
-    precommit_ghost_ = new_precommit_ghost.value();
-
-    if (changed) {
-      SL_TRACE(logger_,
-               "Round #{}: updatePrecommitGhost->true "
-               "(precommit ghost was changed to block {})",
-               round_number_,
-               precommit_ghost_.value());
-      return true;
-    }
+    BOOST_ASSERT(prevote_ghost_.has_value());
+    const auto &prevote_ghost = prevote_ghost_.value();
 
-    SL_TRACE(logger_,
-             "Round #{}: updatePrecommitGhost->false "
-             "(precommit ghost was not changed)",
-             round_number_);
-    return false;
-  }
+    // anything new finalized? finalized blocks are those which have both
+    // 2/3+ prevote and precommit weight.
+    if (precommits_->getTotalWeight() >= threshold_) {
+      auto possible_to_finalize = [&](const VoteWeight &weight) {
+        return weight.total(
+                   VoteType::Precommit, precommit_equivocators_, *voter_set_)
+               >= threshold_;
+      };
 
-  bool VotingRoundImpl::updateCompletable() {
-    // figuring out whether a block can still be committed for is
-    // not straightforward because we have to account for all possible future
-    // equivocations and thus cannot discount weight from validators who
-    // have already voted.
-    const auto tolerated_equivocations = voter_set_->totalWeight() - threshold_;
+      finalized_ = graph_->findAncestor(
+          VoteType::Precommit, prevote_ghost, std::move(possible_to_finalize));
+    };
```

Comment:
redundant ; ?

Reply:
Looks strange. Actual code does not have that.